### ISSUE TITLE
not possible to render (Error during JPEG image loading)

### ISSUE THREAD

**Issue by gHensges** [mozilla/pdf.js#8614]
Link to PDF file (or attach file here): [OBW-OVK.pdf](https://github.com/mozilla/pdf.js/files/1125123/OBW-OVK.pdf)

Configuration:

- Webbrowser and its version: Chrome 59.0.3071.115

- Operatingsystem and its version: Microsoft Windows 7 Enterprise SP1

- PDF.js version: 1.8.516

- Is an extension: no

Steps to reproduce the problem:
1. Load attached pdf-file
2.

What is the expected behavior? (add screenshot)

What went wrong? (add screenshot)
Pdf has been rendered as blank pages. Error trace:

 Error: Invalid XRef stream header
util.js:296     at error (https://mozilla.github.io/pdf.js/build/pdf.worker.js:355:17)
    at XRef_readXRef [as readXRef] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:21322:29)
    at XRef_parse [as parse] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:20992:28)
    at PDFDocument_setup [as setup] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:29900:17)
    at PDFDocument_parse [as parse] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:29789:12)
    at _util.Util.inherit.ensure (https://mozilla.github.io/pdf.js/build/pdf.worker.js:39550:28)
    at Promise (<anonymous>)
    at LocalPdfManager_ensure [as ensure] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:39545:14)
    at LocalPdfManager.BasePdfManager_ensureDoc [as ensureDoc] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:39497:19)
util.js:282 Warning: Indexing all PDF objects
app.js:1056 PDF f5a022ce9c1cfd977eedadde6a39bd36 [1.2 - / -] (PDF.js: 1.8.516)
util.js:282 Warning: Error during JPEG image loading
util.js:282 Warning: Dependent image isn't ready yet
util.js:282 Warning: Error during JPEG image loading
util.js:282 Warning: Dependent image isn't ready yet

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

**Comment by Rob--W** [comment]
**TL;DR: The embedded image has been corrupted, its height value does not represent its actual height.**

---

The referenced file (https://github.com/mozilla/pdf.js/files/1125123/OBW-OVK.pdf) contains a JPEG stream, with the following dictionary:
```
1 0 obj                                                                                                                                                                                                                                                                        
<<                                                                                                                                                                                                                                                                             
/Type /XObject                                                                                                                                                                                                                                                                 
/Subtype /Image                                                                                                                                                                                                                                                                
/Name /Im0                                                                                                                                                                                                                                                                     
/Width 2480                                                                                                                                                                                                                                                                    
/Height 3473                                                                                                                                                                                                                                                                   
/BitsPerComponent 8                                                                                                                                                                                                                                                            
/ColorSpace /DeviceGray
/Filter /DCTDecode
/Length 1080464
>>
stream
...
endstream
endobj
```

However, the referenced image's height is much larger than specified (65535 > 3473). The browser is unable to load the image, and hence the output PDF is blank. Note that Chromium's default PDF viewer is not able to render the PDF either.

I extracted `image.jpeg`, e.g. by opening Chrome's devtools, viewing the network tab and seeing the request to a `blob:`-URL (the image). Then, to download it, run the following from the JavaScript console in the devtools of the same tab:

```
var a = document.createElement('a');
a.download = 'image.jpeg';
// Replace the following with the blob:-URL that is displayed in the devtools
a.href = 'blob:http://localhost:8888/b437bc55-2b13-4401-b526-a248111ad8d4';
a.click();
```

Now, the analysis of the image shows that the height is 65535:

```
$ exiftool image.jpeg 
ExifTool Version Number         : 10.55
File Name                       : image.jpeg
Directory                       : .
File Size                       : 472 kB
...
X Resolution                    : 300
Y Resolution                    : 300
Image Width                     : 2480
Image Height                    : 65535
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8
Color Components                : 1
Image Size                      : 2480x65535
Megapixels                      : 162.5
```

When I open the JPEG file in a hex editor and change the faulty height value FFFF (=65535) to 0D91 (=3473), then the image can be displayed just fine.

To find out which bytes I needed to modify, I read:

- http://vip.sugovica.hu/Sardi/kepnezo/JPEG%20File%20Layout%20and%20Format.htm
- https://en.wikipedia.org/wiki/JPEG#Syntax_and_structure
- http://www.fastgraph.com/help/jpeg_header_format.html
- https://stackoverflow.com/questions/18264357/how-to-get-the-width-height-of-jpeg-file-without-using-library

**Comment by Rob--W** [comment]
The PDF file was blank too in Chromium's PDF viewer and Okular, but shown just fine in Adobe's PDF reader.

A patch for this issue is available at #8628. I am however not 100% sure whether this broken PDF should be supported by PDF.js. Where did you get this PDF file?

**Comment by gHensges** [comment]
I am supporting an application, that fetches e-mails from a server, convert all attachments to multi-tiff, and provide this converted files for further processing. This PDF is coming from somebody unknown to me! I am always confronted with the fact, that customers asks "Adobe Reader can render this file. Why is your program not able to process it" Anyway, thanks for providing this patch!

**Comment by Rob--W** [comment]
Thanks for the info. Let's try to land the patch. None of the existing tests fail, so it should be fine to land it.
